mage_ai/api/resources/BlockResource.py

```diff
@@ -649,6 +649,46 @@ async def create(self, payload, user, **kwargs):
                 replicated_block=replicated_block,
             )
 
+        async def _create_callback(resource):
+            # Replace existing block in pipeline with newly created copy of that block
+            block_uuid_to_remove = payload.get('block_uuid_to_remove')
+            if pipeline and block_uuid_to_remove is not None:
+                callback_blocks = payload.get('callback_blocks')
```

Comment:
You don't need to pass the callback blocks and conditional blocks from the API. You can just get them from the the block of `block_uuid_to_remove` (maybe `block_uuid_to_detach` is a better param name). `block_to_detach = pipeline.get_block(block_uuid_to_detach)`

And I just checked the code, there're actually more associated objects with the block: widget, conditional blocks, callback blocks, extensions, replicated block, variables directory. These need to be associated with the new block. I would suggest adding a method in the Pipeline class to transfer the associated objects.

Here is some code of updating the directories when updating the block uuid. I think they also need to be done for this case:
https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/pipeline.py#L1387-L1393
https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/pipeline.py#L1426-L1441

Reply:
Actually, i think there's a better and simpler way to implement this.
You can just update this method and add an option to detach: https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/block/__init__.py#L2787-L2829
When detach is true, you can just update this line to create the new file instead of rename the file name: https://github.com/mage-ai/mage-ai/blob/master/mage_ai/data_preparation/models/block/__init__.py#L2820C14-L2820C14

Reply:
You can use Block PUT endpoint instead of POST endpoint. Just update the block with the new name and "detach: true".

Reply:
@wangxiaoyou1993 Thank you for the suggestions, made an update to reuse the logic from renaming a block, which does work out better since it handles the associated objects, as well as upstream and downstream dependencies.